Add range-request method

litherum opened the pull request:
This adds `range_request_pfe_method`, which is a method that requests glyph data via range requests. There is one unconditional download, which requests all of the font data that isn't glyph outlines, and then a collection of parallel downloads to download each glyph that is necessary. Once the "base" download has completed, successive page views don't need to download it again. Also, there's a threshold related to network parameters which indicates how hard the method will try to coalesce nearby range requests into a single, longer range request.

There are still some things remaining to do:

- The format of the corpus lists used code points, but this method operates on glyphs. However, it is impossible to translate code points into glyphs with a proper string, so that shaping can occur. Right now, the method just maps the code points via the cmap table, but this should be improved.
- Currently, the test runner requires that the request graphs are insensitive to network parameters. However, a browser can/will be measuring network performance during the first "base" network round-trip. This range-request method is sensitive to network parameters, so these should be exposed to the method.
- I'm using a class in the [font optimizer](https://github.com/litherum/StreamableFonts) Objective-C framework for pulling out the size of each glyph in bytes. However, this isn't playing well with Python's multiprocessing model of `fork()`ing without `exec()`ing (which is explicitly against the rules of Foundation's classes). I should rewrite this function in raw Python.
- HTTP response compression is not adequately modeled. In order to do this right, I'd need the raw bytes of each glyph so that I can gzip them to see what their compression ratio would be. My framework only pulls out the glyph sizes, not their raw data, so I'm emulating gzip by just applying a blanket compression ratio of 46%. This was gathered by just gzipping all the fonts which support Chinese in the Google Fonts repo and finding their average compression ratio.

litherum committed:
Add range-request method

This adds range_request_pfe_method, which is a method that requests glyph data via range requests. There is one unconditional download, which requests all of the font data that isn't glyph outlines, and then a collection of parallel downloads to download each glyph that is necessary. Once the "base" download has completed, successive page views don't need to download it again. Also, there's a threshold related to network parameters which indicates how hard the method will try to coalesce nearby range requests into a single, longer range request.

There are still some things remaining to do:

- The format of the corpus lists used code points, but this method operates on glyphs. However, it is impossible to translate code points into glyphs with a proper string, so that shaping can occur. Right now, the method just maps the code points via the cmap table, but this should be improved.
- Currently, the test runner requires that the request graphs are insensitive to network parameters. However, a browser can/will be measuring network performance during the first "base" network round-trip. This range-request method is sensitive to network parameters, so these should be exposed to the method.
- I'm using a class in the font optimizer Objective-C framework for pulling out the size of each glyph in bytes. However, this isn't playing well with Python's multiprocessing model of `fork()`ing without `exec()`ing (which is explicitly against the rules of Foundation's classes). I should rewrite this function in raw Python.
- HTTP response compression is not adequately modeled. In order to do this right, I'd need the raw bytes of each glyph so that I can gzip them to see what their compression ratio would be. My framework only pulls out the glyph sizes, not their raw data, so I'm emulating gzip by just applying a blanket compression ratio of 46%. This was gathered by just gzipping all the fonts which support Chinese in the Google Fonts repo and finding their average compression ratio.

garretrieger commented on analysis/analyzer.py:
```diff
@@ -187,6 +195,24 @@ def read_text_input(input_data_path):
     text_format.Merge(input_data_file.read(), data_set)
   return data_set
 
+def read_json_input(input_data_path):
```
Could you add a comment documenting the expected format of the json.

garretrieger commented on analysis/pfe_methods/range_request_pfe_method.py:
```diff
@@ -0,0 +1,206 @@
+    self.loaded_glyphs = defaultdict(set)
+    # FIXME: Build this framework as part of the project.
+    bundle = objc.loadBundle("Optimizer", globals(), bundle_path="/Users/mmaxfield/Build/Products/Debug/Optimizer.framework")
+    self.GlyphSizeComputer = bundle.classNamed_("ExposedGlyphSizeComputer")
```
FYI you can use fontTools to gather the size of glyphs. Something like this:

for glyphname, glyph in font["glyf"].glyphs.items():
  len(glyph.data)

garretrieger commented on analysis/pfe_methods/range_request_pfe_method.py:
```diff
@@ -0,0 +1,206 @@
+  def get_base_font_size(self, font_sizes):
+    return font_sizes.fontSize() - sum(font_sizes.glyphSizes())
+
+  class GlyphRange:
```
Recommend using namedtuples (https://docs.python.org/3/library/collections.html#collections.namedtuple) for value classes. Less boilerplate is needed.

svgeesus replied:
Because this looks like a coding style choice and has sat here unanswered for a while, I think it best to merge  as-is because it is more valuable to get some analysis results frm this PFE method than to worry about longer term code maintainability

litherum replied:
> Named tuples are immutable

https://stackoverflow.com/a/19257700

litherum committed:
Removes objective-c dependency in range-request method

Also add in proper zlib support, so we don't have to
guess what the compression ratio would be.

litherum committed:
Various fixes

- Fixing Ahem optimized font. The font wasn't optimized correctly.
- Addressing comments in https://github.com/w3c/PFE-analysis/pull/35
- Fixing the build and tests

litherum committed:
Various fixes

The analysis of the full data set now runs on the range request method.

garretrieger commented on patch_subset/py/patch_subset_method.py:
```diff
@@ -84,8 +84,10 @@ def __init__(self, font_loader, font_id, page_view_count,
                with_codepoint_remapping):
     font_directory_c = c_char_p(font_loader.directory().encode("utf-8"))
     font_id_c = c_char_p(font_id.encode("utf-8"))
+    default_font = font_loader.default_font() or ""
```
Instead of adding default_font support to the C++ code, you could check if font_id is empty or None here and then swap in the default font before calling new_session.

svgeesus replied:
Not clear this is a blocking comment either. I prefer to merge now and fix later if needed.

svgeesus commented:
So this PR says approved but  also see a request for two changes. Should this be merged, or discussed first, or  altered first?

svgeesus commented:
Looks like a rebase would fix this (and perhaps running the formatting script)

litherum commented:
Landed in https://github.com/w3c/PFE-analysis/commit/7efa46945c690f84bcd48848cbfd8bcba638b2a4
